docker/test_docker.py

```diff
@@ -0,0 +1,589 @@
+
+import httpx
+import pypiserver
+import pytest
+
+
+THIS_DIR = Path(__file__).parent
+ROOT_DIR = THIS_DIR.parent
+DOCKERFILE = ROOT_DIR / "Dockerfile"
+FIXTURES = ROOT_DIR / "fixtures"
+MYPKG_ROOT = FIXTURES / "mypkg"
+HTPASS_FILE = FIXTURES / "htpasswd.a.a"
+
+
+# This is largely useless when using pytest because of the need to use the name
+# of the fixture as an argument to the test function or fixture using it
+# pylint: disable=redefined-outer-name
+#
+# Also useless for our test context, where we may want to group test functions
+# in a class to share common fixtures, but where we don't care about the
+# `self` instance.
+# pylint: disable=no-self-use
+
+
+@pytest.fixture(scope="session")
+def image() -> str:
+    """Build the docker image for pypiserver.
+
+    Return the tag.
+    """
+    tag = "pypiserver:test"
+    proc = subprocess.Popen(
+        ("docker", "build", "--file", str(DOCKERFILE), "--tag", tag, ROOT_DIR),
+        cwd=ROOT_DIR,
+    )
+    proc.communicate()
+    assert proc.returncode == 0
+    return tag
+
+
+@pytest.fixture(scope="session")
+def mypkg_build() -> None:
+    """Ensure the mypkg test fixture package is build."""
+    # Use make for this so that it will skip the build step if it's not needed
+    proc = subprocess.Popen(("make", "mypkg"), cwd=ROOT_DIR)
+    proc.communicate()
+    assert proc.returncode == 0
+
+
+@pytest.fixture(scope="session")
+def mypkg_paths(
+    mypkg_build: None,  # pylint: disable=unused-argument
+) -> t.Dict[str, Path]:
+    """The path to the mypkg sdist file."""
+    dist_dir = Path(MYPKG_ROOT) / "dist"
+    assert dist_dir.exists()
+
+    sdist = dist_dir / "pypiserver_mypkg-1.0.0.tar.gz"
+    assert sdist.exists()
+
+    wheel = dist_dir / "pypiserver_mypkg-1.0.0-py2.py3-none-any.whl"
+    assert wheel.exists()
+
+    return {
+        "dist_dir": dist_dir,
+        "sdist": sdist,
+        "wheel": wheel,
+    }
+
+
+def wait_for_container(port: int) -> None:
+    """Wait for the container to be available."""
+    for _ in range(60):
+        try:
+            httpx.get(f"http://localhost:{port}").raise_for_status()
+        except (httpx.RequestError, httpx.HTTPStatusError):
+            time.sleep(1)
+        else:
+            return
+
+    # If we reach here, we've tried 20 times without success.
+    raise RuntimeError("Could not connect to pypiserver container")
+
+
+def uninstall_pkgs() -> None:
+    """Uninstall any packages we've installed."""
+    proc = subprocess.Popen(
+        ("pip", "freeze"), stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    )
+    out, err = proc.communicate()
+    assert proc.returncode == 0, err and f"{err.decode()}"
+    if any(
+        ln.strip().startswith("pypiserver-mypkg")
+        for ln in out.decode().splitlines()
+    ):
+        proc = subprocess.Popen(("pip", "uninstall", "-y", "pypiserver-mypkg"))
+        proc.communicate()
+
+
+@pytest.fixture(scope="session")
+def session_cleanup() -> t.Iterator[None]:
```

Comment:
It looks like this fixture is never used

Reply:
Ah I meant to make it autouse and forgot to. Another good catch